frame/contracts/src/lib.rs

```diff
@@ -967,6 +989,156 @@ struct InternalOutput<T: Config, O> {
 	result: Result<O, ExecError>,
 }
 
+/// Helper trait to wrap contract execution entry points into a signle function
+/// [`Invokable::run_guarded`].
+trait Invokable<T: Config, O> {
+	/// Single entry point to contract execution.
+	/// Downstream execution flow is branched by implementations of [`Invokable`] trait:
+	///
+	/// - [`InstantiateInput::run`] runs contract instantiation,
+	/// - [`CallInput::run`] runs contract call.
+	///
+	/// We enforce a re-entrancy guard here by initializing and checking a boolean flag through a
+	/// global reference.
+	fn run_guarded(&self, context: ExecContext<T>) -> InternalOutput<T, O> {
```

Comment:
It might be cleaner to just make `ExecContext` a field of both `CallInput` and `InstantiateInput`.

Reply:
This would add complexity to the code: need to add lifetime params to the types (all because of the `debug_message` field), and a getter func for the `context` to the trait. On the other hand, separating the input of the call (like business data) from its execution context (like gas_limit, attached payment, etc.) seems to be a right thing to do. That's being said, moving `data` field out of context to the input really makes sense. I've done just that.

Reply:
> and a getter func for the context to the trait.

Yeah forgot about this. Let's leave it as is.

> That's being said, moving data field out of context to the input really makes sense. I've done just that.

Why? We just duplicated a common field by doing so.